alarm_active_table.cpp

```diff
@@ -0,0 +1,423 @@
+      // Convert the index string OID
+      int model_pointer_len = alarmModelTable_oid_len + entry_column_oid_len + 3;
+      // Append the index array on to the Alarm Model Table OID, inserting
+      // the OID ".1.3." for the entry in the first non index column
+      model_pointer = new oid[model_pointer_len];
+      copy(alarmModelTable_oid, alarmModelTable_oid + alarmModelTable_oid_len, model_pointer);
+
+      copy(entry_column_oid, entry_column_oid + entry_column_oid_len, model_pointer + alarmModelTable_oid_len);
+      
+      model_pointer[alarmModelTable_oid_len + entry_column_oid_len] = 0;
+      model_pointer[alarmModelTable_oid_len + entry_column_oid_len + 1] = context->_alarm_table_def->alarm_index(); 
+      model_pointer[alarmModelTable_oid_len + entry_column_oid_len + 2] = context->_alarm_table_def->severity();      
+      
+      snmp_set_var_typed_value(var, ASN_OBJECT_ID,
+                               (u_char*) model_pointer,
+                               model_pointer_len * sizeof(oid));
+
+    }
+    break;
+    
+    case COLUMN_ALARMACTIVESPECIFICPOINTER:
+    {
+      // This would point to the model-specific active alarm list
+      // for this alarm. We do not support model-specific Alarm-MIB
+      // and hence this will always be the 0.0 OID
+      snmp_set_var_typed_value(var, ASN_OBJECT_ID,
+                               (u_char*) zero_dot_zero_oid,
+                               sizeof(zero_dot_zero_oid));
+    }
+    break;
+    
+    
+    default: /** We shouldn't get here */
+    {
+      TRC_ERROR("unknown column: alarmActiveTable_get_value");
+      return SNMP_ERR_GENERR;
+    }
+  }
+
+  return SNMP_ERR_NOERROR;
+}
+
+/************************************************************
+ * 
+ *  Create a new row context and initialize its oid index.
+ */
+void alarmActiveTable_create_row(char* name,
+                                 alarmActiveTable_SNMPDateTime* datetime,
+                                 unsigned long index,
+                                 AlarmTableDef& def)
+{
+  alarmActiveTable_context* ctx = SNMP_MALLOC_TYPEDEF(alarmActiveTable_context);
+  ctx->_alarm_table_def = &def;
+  if (pointers[ctx->_alarm_table_def->alarm_index()])
+  {
+    if (ctx->_alarm_table_def->severity() == pointers[ctx->_alarm_table_def->alarm_index()]->_alarm_table_def->severity())
+    {
+      return;
+    }
+  }
+  if (!ctx)
+  {
+    TRC_ERROR("malloc failed: alarmActiveTable_create_row");
+    return;
+  }
+        
+  if (alarmActiveTable_index_to_oid(name, datetime, index, &ctx->_index) != SNMP_ERR_NOERROR)
+  {
+    if (ctx->_index.oids != NULL)
+    {
+      free(ctx->_index.oids);
+    }
+
+    free(ctx);
+    return;
+  }
+  if (ctx->_alarm_table_def->severity() == 1)
+  {
+    alarmActiveTable_delete_row(def);
+  }
+  else
+  {
+    if (pointers[ctx->_alarm_table_def->alarm_index()])
+    {
+      CONTAINER_REMOVE(cb.container, pointers[ctx->_alarm_table_def->alarm_index()]);
```

Comment:
This feels like it makes more sense above (alongside your other check for the existing row).